### PR TITLE
[megatron] fix: correct CP>1 batch_num_tokens aggregation in engine path

### LINKED ISSUE #5983: [megatron][engine][context parallel] batch_num_tokens aggregation undercounts valid tokens when CP > 1
### Summary

In the Megatron engine path, `batch_num_tokens` appears to be aggregated over the pure DP group in `transformer_impl.py`.

When `CP > 1`, each CP rank only holds a 1/CP slice of the sequence, so reducing `batch_num_tokens` over pure DP undercounts valid tokens by a factor of `CP`.

This affects `loss_agg_mode="token-mean"` normalization in the engine path.

### Current behavior

For `CP > 1`, `batch_num_tokens` is effectively smaller than the true valid-token count for the full DP×CP shard.

As a result, token-mean normalization is incorrect.

### Expected behavior

For `CP > 1`, `batch_num_tokens` should be reduced over the DP×CP group, while `dp_size` should remain pure DP.

### Proposed fix

Use `mpu.get_data_parallel_group(with_context_parallel=True)` when reducing `batch_num_tokens` in the Megatron engine path.

### Validation

I have:
- a minimal code fix in `transformer_impl.py`
- a focused test / reproducer for the CP>1 case
- manual smoke coverage for the engine path

### Related context

Related to #4852  
Potentially related to #1332

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The bug can be reproduced with a minimal Megatron engine-path setup where `context_parallel_size > 1`.

#### Config

- `tensor_model_parallel_size = 1`
- `pipeline_model_parallel_size = 1`
- `context_parallel_size = 2`
- `data_parallel_size = 1` (minimal case)
- `loss_agg_mode = "token-mean"`

#### Minimal reproducer

```python
import torch
import torch.distributed as dist
from megatron.core import parallel_state as mpu

def main():
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    local_rank = int(torch.cuda.current_device())
    device = torch.device("cuda", local_rank)

    # Assume Megatron parallel state has already been initialized with:
    # TP=1, PP=1, CP=2
    #
    # Each CP rank only sees a 1/CP slice of the sequence.
    # For a fully valid local slice:
    loss_mask = torch.ones(8, device=device)  # 8 valid tokens on each CP rank
    local_tokens = loss_mask.sum().to(torch.int64)  # = 8 on each rank

    # Current behavior in the buggy engine path:
    wrong = local_tokens.clone()
    dist.all_reduce(
        wrong,
        op=dist.ReduceOp.SUM,
        group=mpu.get_data_parallel_group(),   # pure DP group
    )

    # Expected behavior:
    correct = local_tokens.clone()
    dist.all_reduce(
        correct,
        op=dist.ReduceOp.SUM,
        group=mpu.get_data_parallel_group(with_context_parallel=True),  # DP×CP group
    )

    if rank == 0:
        print(f"local_tokens={int(local_tokens)}")
        print(f"wrong_batch_num_tokens={int(wrong)}")
        print(f"correct_batch_num_tokens={int(correct)}")

        # With CP=2, wrong_batch_num_tokens is smaller by a factor of 2.
        # This undercounts valid tokens and makes token-mean normalization incorrect.

if __name__ == "__main__":
    main()
```

### Expected behavior

For `CP > 1`, the Megatron engine path should aggregate `batch_num_tokens` over the correct DP×CP group, while keeping `dp_size` equal to pure DP. This should make `loss_agg_mode="token-mean"` use the correct valid-token count and avoid CP-dependent undercounting.

### PR DESCRIPTION
**Pull request by alex27594**
## Summary

This PR fixes a correctness bug in the **Megatron engine path** when `context_parallel_size > 1`.

In `MegatronEngine.forward_backward_batch()`, `batch_num_tokens` is used as the denominator for `loss_agg_mode="token-mean"` normalization. When context parallelism is enabled, each CP rank only owns a **1/CP slice of the sequence**, so `batch_num_tokens` must be aggregated over the **DP×CP group**, not the pure DP group.

Using the pure DP group undercounts valid tokens by a factor of `CP`, which makes token-mean loss normalization incorrect in the engine path for `CP > 1`.

This PR fixes that by reducing `batch_num_tokens` over:

```python
mpu.get_data_parallel_group(with_context_parallel=True)
```

while intentionally keeping `dp_size` equal to **pure DP**.

Closes #5983 
Related to #4852  
Potentially related to #1332

---

## Problem

The Megatron engine path computes:
- `batch_num_tokens`
- `dp_size`

and passes them into downstream loss aggregation.

For `loss_agg_mode="token-mean"`, the intended semantics are:

- `batch_num_tokens` = true valid-token count for the full shard participating in the token-mean denominator
- `dp_size` = pure data parallel size

This is correct for standard DP/TP/PP layouts. However, when `context_parallel_size > 1`, the sequence is partitioned across CP ranks, so each rank only sees a fraction of the valid tokens.

If `batch_num_tokens` is reduced over the **pure DP group**, the token count is smaller than the true global token count by a factor of `CP`. That leads to incorrect normalization and over-scaled loss values.

In practice, for `CP > 1`, the current behavior effectively computes:

```text
batch_num_tokens ≈ true_tokens / CP
```

which makes `token-mean` normalization wrong by construction.

---

## Root cause

The denominator used for token-mean aggregation is gathered over the wrong communication group.

For token counting:
- **incorrect** group: pure DP
- **correct** group: DP×CP

This distinction matters because CP partitions the sequence itself. Each CP rank contributes only a partial token view, so token counting must include the context-parallel dimension.

At the same time, `dp_size` must remain **pure DP**, because `agg_loss(..., loss_agg_mode="token-mean")` expects DP scaling semantics there, not DP×CP scaling.

So the correct combination is:
- `batch_num_tokens`: reduce over **DP×CP**
- `dp_size`: keep as **pure DP**

---

## Fix

This PR changes the `batch_num_tokens` all-reduce in `MegatronEngine.forward_backward_batch()` from the pure DP group to the CP-aware DP×CP group:

```python
torch.distributed.all_reduce(
    batch_num_tokens,
    op=torch.distributed.ReduceOp.SUM,
    group=mpu.get_data_parallel_group(with_context_parallel=True),
)
```

and keeps:

```python
dp_size = self.get_data_parallel_size()
```

unchanged.

This is intentionally a **minimal production change**:
- no refactors
- no algorithm changes
- no changes to non-CP paths
- no changes to user-facing configuration

It only corrects the communication group used for token counting.

---

## Why this fix is correct

When `CP > 1`, each CP rank holds only a partial sequence slice. That means the true token-mean denominator must include all valid tokens across the CP dimension.

Reducing `batch_num_tokens` over DP×CP restores exactly that behavior.

The fix also preserves the existing and intended `agg_loss(...)` contract:
- `batch_num_tokens` reflects the full valid-token count
- `dp_size` reflects pure DP only

That makes the Megatron engine path consistent with the expected semantics of `loss_agg_mode="token-mean"`.

---

## Validation

This PR adds focused distributed coverage for the CP>1 case.

### `tests/workers/engine/test_special_megatron_engine_cp_loss_norm.py`

This test is intentionally narrow and checks three things:

1. **DP×CP token aggregation is the correct denominator**
   - verifies that all-reducing over DP×CP yields the true global token count
   - verifies that pure-DP reduction is too small when `CP > 1`

2. **`agg_loss(..., loss_agg_mode="token-mean")` behaves correctly once the denominator is correct**
   - validates the expected normalization math directly

3. **The engine path uses the CP-aware group in `forward_backward_batch()`**
   - acts as a regression marker for the exact CP-specific behavior fixed in this PR

### `tests/workers/engine/run_megatron_engine_cp_loss_norm.sh`

A simple manual smoke script is included to make the reproducer easy to run under `torchrun`.

The script is intentionally lightweight and targeted at this exact bug, so reviewers and future contributors can validate the behavior without needing a full training recipe or model checkpoint.

---

## Why this PR is useful for upstream

This is a **small correctness fix** in an actively maintained path:
- it affects the **modern Megatron engine path**
- it is directly relevant to users running with `context_parallel_size > 1`
- it improves loss-normalization correctness rather than changing behavior for unrelated configurations

It also complements existing engine-path normalization discussions such as #4852 by adding a focused fix and targeted coverage for a specific CP-related failure mode.

In short, this PR:
- fixes a real normalization bug
- keeps the production diff minimal
- adds narrowly scoped regression coverage for the exact issue being fixed

---

## Scope

This PR is intentionally limited to:
- the Megatron **engine path**
- `CP > 1` token-count normalization
- focused distributed validation

It does **not** include:
- legacy worker changes
- FSDP changes
- rollout changes
- unrelated refactors
- broad Megatron behavior changes outside the token-count aggregation path

---

## Local validation

The following were used during validation:

```bash
pre-commit run --all-files --show-diff-on-failure --color=always

# manual smoke
bash tests/workers/engine/run_megatron_engine_cp_loss_norm.sh

# direct distributed run
torchrun --standalone --nnodes=1 --nproc_per_node=8 \
  tests/workers/engine/test_special_megatron_engine_cp_loss_norm.py
```

I can narrow or reshape the validation further if maintainers prefer a different test style for engine-path special tests.

### CODE REVIEW

**Comment by wuxibin89** [comment]
Megatron will scale loss by `cp_group_size`, so `batch_num_tokens` should be aggregated over DP group instead of DPxCP group?
https://github.com/NVIDIA/Megatron-LM/blob/core_v0.16.0/megatron/core/pipeline_parallel/schedules.py#L268
